nlprule/src/compile/impls.rs

```diff
@@ -33,6 +37,201 @@ use crate::{
 
 use super::{parse_structure::BuildInfo, Error};
 
+impl Spell {
+    fn new(
+        fst: Vec<u8>,
+        multiwords: DefaultHashMap<String, Vec<(Vec<String>, SpellInt)>>,
+        max_freq: usize,
+        map: DefaultHashMap<String, String>,
+        lang_options: SpellLangOptions,
+    ) -> Self {
+        let mut spell = Spell {
+            fst,
+            multiwords,
+            max_freq,
+            map,
+            lang_options,
+            ..Default::default()
+        };
+        spell.ingest_options();
+        spell
+    }
+
+    #[allow(clippy::clippy::too_many_arguments)] // lots of arguments here but not easily avoidable
+    fn add_line(
+        word: &str,
+        freq: usize,
+        words: &mut DefaultHashMap<String, SpellInt>,
+        multiwords: &mut DefaultHashMap<String, Vec<(Vec<String>, SpellInt)>>,
+        variant_index: usize,
+        // in some LT lists an underline denotes a prefix such that e.g. "hin_reiten" also adds "hingeritten"
+        underline_denotes_prefix: bool,
+        build_info: &mut BuildInfo,
+        tokenizer: &Tokenizer,
+    ) {
+        let tokens = tokenizer.get_token_strs(word);
+
+        if tokens.len() > 1 {
+            assert!(!(underline_denotes_prefix && word.contains('_'))); // not supported in entries spanning multiple tokens
+
+            let mut int = SpellInt::default();
+            int.add_variant(variant_index);
+
+            // we do not add the frequency for multiwords - they are not used for suggestions, just to check validity
+
+            multiwords
+                .entry(tokens[0].to_owned())
+                .or_insert_with(Vec::new)
+                .push((
+                    tokens[1..]
+                        .iter()
+                        .filter(|x| !x.trim().is_empty())
+                        .map(|x| (*x).to_owned())
+                        .collect(),
+                    int,
+                ));
+        } else if word.contains('_') && underline_denotes_prefix {
+            assert!(!word.contains('\\')); // escaped underlines are not supported
+            let mut parts = word.split('_');
+
+            let prefix = parts.next().unwrap();
+            let suffix = parts.next().unwrap();
```

Comment:
I'd strongly suggest to use `debug_assert_eq(parts.clone().count(), )` and `.expect()` instead.